R/pkg_metric_has_news.R

```diff
@@ -0,0 +1,99 @@
+#' Assess a package for the presence of a NEWS file
+#'
+#' @param x a packge reference object
+#' @param ... additional arguments unused
+#'
+#' @return a \code{pkg_metric} object
+#' @export
+#'
+assess_has_news <- function(x, ...) {
+  UseMethod("assess_has_news")
+}
+
+
+
+#' @rdname assess_has_news
+#' @export
+#' @importFrom xml2 xml_find_all xml_attrs
+assess_has_news.pkg_remote <- function(x, ...) {
+  # scrape CRAN html for NEWS links
+  NEWS_links <- xml2::xml_find_all(x$web_html, xpath = '//a[.="NEWS"]')
+
+  # add NEWS link url metadata to package environment
+  x$materials_urls <- sprintf("%s/%s",
+    x$web_url,
+    vapply(xml2::xml_attrs(NEWS_links), "[", character(1L), "href"))
+
+  news_result <- rep(list(content = NULL, valid = TRUE), length(x$materials_urls))
+  structure(
+    news_result,
+    class = c("pkg_metric_has_news", "pkg_metric", class(news_result)))
+}
+
+
+
+#' @rdname assess_has_news
+#' @export
+assess_has_news.pkg_install <- function(x, ...) {
+  news_result <- assess_news_in_dir(system.file(package = x$name), ...)
+  structure(
+    news_result,
+    class = c("pkg_metric_has_news", "pkg_metric", class(news_result)))
+}
+
+
+
+#' @rdname assess_has_news
+#' @export
+assess_has_news.pkg_source <- function(x, ...) {
+  news_result <- assess_news_in_dir(x$path, ...)
+  structure(
+    news_result,
+    class = c("pkg_metric_has_news", "pkg_metric", class(news_result)))
+}
+
+
+
+#' @importFrom tools file_ext
+assess_news_in_dir <- function(x) {
+  news <- list()
+
+  # accommodate NEWS.Rd, NEWS.md, etc
+  news$files <- list.files(x, pattern = "^NEWS\\.*")
```

Comment:
It might be a feature or a bug in `list.files`. the output has been sort instead of matching the input package. You can see from below, `rtf` package (5th pacakge) have filename "NEWS", but it is in the first place in the output. 

```
> list.files(pkg$path[1:5], pattern = "^NEWS\\.*", full.names = TRUE)
[1] "/SFS/user/ctc/zhanyilo/R/covr/NEWS.md"    "/SFS/user/ctc/zhanyilo/R/gt/NEWS.md"     
[3] "/SFS/user/ctc/zhanyilo/R/janitor/NEWS.md" "/SFS/user/ctc/zhanyilo/R/rex/NEWS.md"    
[5] "/SFS/user/ctc/zhanyilo/R/rtf/NEWS"       
> list.files(pkg$path[1:5], pattern = "^NEWS\\.*")
[1] "NEWS"    "NEWS.md" "NEWS.md" "NEWS.md" "NEWS.md"
> pkg$path[1:5]
[1] "/SFS/user/ctc/zhanyilo/R/covr"    "/SFS/user/ctc/zhanyilo/R/gt"      "/SFS/user/ctc/zhanyilo/R/janitor"
[4] "/SFS/user/ctc/zhanyilo/R/rex"     "/SFS/user/ctc/zhanyilo/R/rtf"   
``` 

* in addition, is it possible to have `news` or `news.md` ? should we use `toupper` here?

Reply:
My intention here was that the `assess_*` functions would be called only with a `pkg_reference` object as input, such that `x` would be a single package directory, such that `list.files(...)` would give all available NEWS files from the package. 

For the time being, I think NEWS must be capitalized (per the extensions manual, it also [must be a `NEWS.Rd` file](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#index-news), but I believe CRAN accepts a `NEWS.md` file these days it it seems to be the growing preference). Given how flexible the format is here, I wouldn't be surprised if a lowercase version was possible in the future, but I'd prefer not to try to address it unless that becomes a community standard.

Reply:
I see. Could we add a checking to ensure the length of `x` is 1 if it is the intention.

Reply:
I'd prefer not to add a check around it for now. Given that this is being used in a non-exported function and is only called internally using paths extracted from a `pkg_ref` object, I don't see it as a big threat. If we ever think about vectorizing `assess_*` functions, there would be a number of places that would have to be modified upstream of this function so it would be unavoidable to do some modifications anyway.

Reply:
@dgkf , I agree. Thanks for the clarification. I think we can complete the review after passing the Travis CI check.